### PR TITLE
SOL-112342: Added new request-reply blocking and non-blocking samples

### PR DESCRIPTION
**Pull request by oodigie**
The goal is to add request-reply blocking and non-blocking samples to demonstrate the use of the request-reply feature in the Solace PubSub+ Go APIs.

**Changes:**
- feat[SOL-112342]: Added new request-reply blocking and non-blocking samples to demonstrate the feature using the Solace PubSub+ Go APIs
- chore: bumped version of the `solace/dev/go/messaging` package to 1.6.0 which has support for request-reply

### CODE REVIEW

**Comment by oodigie** [comment]
As the Request-Reply feature for the PubSub+ Go APIs has not be released, this would be a draft-pull-request so it is not accidentally merged before then.

File: patterns/request-reply/direct_replier_blocking.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	"os"
+	"os/signal"
+	"time"
+
+	"solace.dev/go/messaging"
+	"solace.dev/go/messaging/pkg/solace/config"
+	"solace.dev/go/messaging/pkg/solace/resource"
+)
+
+// getEnvRPNB function
```
Comment (TamimiGitHub) [inline patterns/request-reply/direct_replier_blocking.go:14]: What is `RPNB`?

Comment (oodigie) [reply patterns/request-reply/direct_replier_blocking.go:14]: I forgot that code comment there from my testing, sorry about that. I will remove it and update the PR now. Thanks!

**Comment by aaron-613** [comment]
When running the non-blocking Requestor, if the Replier is terminated early such that there are outstanding request messages, then I get this when I `Ctrl+C` on the Requestor:
```
Published message with sequence number: 6 on topic: solace/samples/go/request-reply
Published message with sequence number: 7 on topic: solace/samples/go/request-reply
^C
Request-Reply Publisher Terminated?  true
panic: publisher has been terminated, no reply messages to receive

goroutine 70 [running]:
main.ReplyMessageHandler(0x0, 0x0, 0x0, 0x0, 0x7fea7d0250e0, 0xc00013c160)
        /home/alee/GitHub/review/solace-samples-go/patterns/request-reply/direct_requestor_non_blocking.go:36 +0x1ca
solace.dev/go/messaging/internal/impl/publisher.(*requestReplyMessagePublisherImpl).createReplyCorrelation.func1(0x68d7d0, 0xc000098590, 0x12a05f200, 0x0)
        /home/alee/go/pkg/mod/solace.dev/go/messaging@v1.6.1/internal/impl/publisher/request_reply_message_publisher_impl.go:855 +0x209
created by solace.dev/go/messaging/internal/impl/publisher.(*requestReplyMessagePublisherImpl).Publish
        /home/alee/go/pkg/mod/solace.dev/go/messaging@v1.6.1/internal/impl/publisher/request_reply_message_publisher_impl.go:489 +0x149
exit status 2
```

To repro: start replier (either one), start non-blocking requestor, stop replier, terminate requestor after a few seconds.

**Comment by aaron-613** [comment]
Can we change the topic that the apps are using?  This is to align with other APIs (Java, etc.), so that they can be used interchangeably (i.e. JCSMP requestor, Python replier).

Change requestor topic to: `solace/samples/go/request-reply`
Change replier to listen to: `solace/samples/*/request-reply`

(I also just noticed that Python is wrong, so I'll raise an Issue for that)

**Comment by aaron-613** [comment]
When running the blocking replier, I get another panic as soon as I hit `Ctrl+C`, regardless if I receive messages or not:
```
$ go run direct_replier_blocking.go
Connected to the broker?  true
Request-Reply Receiver running?  true

===Interrupt (CTR+C) to handle graceful terminaltion of the subscriber===
Message Topic Subscription:  solace/samples/go/request-reply
^C
Request-Reply Receiver Terminated?  true
panic: receiver has been terminated, no messages to receive

goroutine 36 [running]:
main.main.func1(0x6b65c0, 0xc0000ac078, 0x6b6a00, 0xc0000e00c0, 0xc000090530)
        /home/alee/GitHub/review/solace-samples-go/patterns/request-reply/direct_replier_blocking.go:83 +0x5b4
created by main.main
        /home/alee/GitHub/review/solace-samples-go/patterns/request-reply/direct_replier_blocking.go:75 +0x6fb
exit status 2

**Comment by oodigie** [comment]
@aaron-613, I have responded to and fixed the issues you raised. Thanks!

**Comment by aaron-613** [comment]
Hi @oodigie , big apologies but I got the topic format/structure I requested wrong... cut-and-paste error!  To match other similar samples, could it look something like:

Publisher: `solace/samples/go/direct/request`
Subscriber: `solace/samples/*/direct/request`

That should make it interoperate with, for example, JCSMP samples.

One last thing: if the non-blocking requestor is running, and does not receive a response, it prints out:
```
The reply timed out with timed out waiting for reply message for request publish with user context : %!s(<nil>)
```

The "user context" printed out at the end looks like a string substitution gone wrong?  I'm wondering if that is supposed to show something else?  Or maybe needs to be removed..?  Or maybe show which # number request message timed out?

Thanks!

**Comment by oodigie** [comment]
@aaron-613, I will make the updates. As for the "user context" string substitution in the reply timeout error message, I will update this too since we do not pass any user context for these samples. I should push the updates in the next few minutes

**Comment by oodigie** [comment]
@aaron-613, please I have made the suggested changes. Kindly review again, thanks!

**Comment by aaron-613** [comment]
Yup!  Tested interoperability with JCSMP, all console output looks good now.  I'll have @TamimiGitHub do a quick code spot-check then I'll merge.

File: patterns/request-reply/direct_replier_blocking.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	fmt.Println("Message Topic Subscription: ", topicSubscription.GetName())
+
+	// have receiver push request messages to request message handler
+	// Run in a seperate Go routine
+	var receivedMsgCounter = 0 // counter for the received messages
+
+	// Run forever until an interrupt signal is received
+	for requestReplyReceiver.IsRunning() {
+		// have receiver push request messages to request message handler
+		message, replier, regErr := requestReplyReceiver.ReceiveMessage(-1)
```
Comment (TamimiGitHub) [inline patterns/request-reply/direct_replier_blocking.go:80]: Can we please put a comment or a reference to the API doc to why `-1` is added as the function parameter to this ReceiveMessage? Thanks!

**Review by TamimiGitHub — APPROVED** [review]
reviewed the files and ran locally. LGTM!

File: patterns/request-reply/direct_replier_blocking.go (hunk 2 above)
Comment (oodigie) [reply patterns/request-reply/direct_replier_blocking.go:80]: @TamimiGitHub, this is the reference to the API doc - https://pkg.go.dev/solace.dev/go/messaging@v1.6.1/pkg/solace#RequestReplyMessageReceiver . I can add it in a separate PR if you want me to

Comment (TamimiGitHub) [reply patterns/request-reply/direct_replier_blocking.go:80]: Could you add it to the same PR please and we can merge right away?

**Comment by oodigie** [comment]
@aaron-613 and @TamimiGitHub, please can I go ahead and merge this PR since it has been approved?

**Comment by TamimiGitHub** [comment]
PR merged 👍

File: patterns/request-reply/direct_replier_blocking.go (hunk 2 above)
Comment (oodigie) [reply patterns/request-reply/direct_replier_blocking.go:80]: I have added it in the provision/deprovision PR - #13

**Pull request merged**